src/algorithms/tracking/TrackPropagation.cc

```diff
@@ -67,66 +67,82 @@ namespace eicrecon {
             if(!result) continue;
 
             // Add to output collection
-            track_points.push_back(result);
+            track_points.push_back(std::move(result));
         }
 
         return track_points;
     }
 
 
 
-    edm4eic::TrackSegment* TrackPropagation::propagateToSurfaceList(const eicrecon::TrackingResultTrajectory *traj,
-                                                                    std::vector<std::shared_ptr<Acts::Surface>> targetSurfaces) {
-      // start a mutable TrackSegment
-      edm4eic::MutableTrackSegment track_segment;
-      decltype(edm4eic::TrackSegmentData::length)      length       = 0;
-      decltype(edm4eic::TrackSegmentData::lengthError) length_error = 0;
-
-      // loop over projection-target surfaces
-      for(const auto& targetSurf : targetSurfaces) {
-
-        // project the trajectory `traj` to this surface
-        edm4eic::TrackPoint *point;
-        try {
-          point = propagate(traj, targetSurf);
-        } catch(std::exception &e) {
-          m_log->warn("<> Exception in TrackPropagation::propagateToSurfaceList: {}; skip this TrackPoint and surface", e.what());
-        }
-        if(!point) {
-          m_log->trace("<> Failed to propagate trajectory to this plane");
-          continue;
-        }
-
-        // logging
-        m_log->trace("<> trajectory: x=( {:>10.2f} {:>10.2f} {:>10.2f} )",
-            point->position.x, point->position.y, point->position.z);
-        m_log->trace("               p=( {:>10.2f} {:>10.2f} {:>10.2f} )",
-            point->momentum.x, point->momentum.y, point->momentum.z);
-
-        // update the `TrackSegment` length
-        // FIXME: `length` and `length_error` are currently not used by any callers, and may not be correctly calculated here
-        if(track_segment.points_size()>0) {
-          auto pos0 = point->position;
-          auto pos1 = std::prev(track_segment.points_end())->position;
-          auto dist = edm4eic::magnitude(pos0-pos1);
-          length += dist;
-          m_log->trace("               dist to previous point: {}", dist);
-        }
-
-        // add the `TrackPoint` to the `TrackSegment`
-        track_segment.addToPoints(*point);
-
-      } // end `targetSurfaces` loop
-
-      // output
-      track_segment.setLength(length);
-      track_segment.setLengthError(length_error);
-      return new edm4eic::TrackSegment(track_segment);
+    std::unique_ptr<edm4eic::TrackSegmentCollection> TrackPropagation::propagateToSurfaceList(
```

Comment:
If you need to return a collection, it probably should be one of the STL ones, not PODIO.

Reply:
We should have been working with PODIO collections from the start. The `std::vector<raw pointer to PODIO object>` object is not easy to work with in practice, because the PODIO collection which owns them is never clear (ownership is needed to make relations); this initial aversion of PODIO collections has been the cause of multiple memory issues downstream.

Reply:
Algorithm returns STL collections of owned smart pointers, factory packages it for a specific way to output information by moving pointers out into the PODIO collection.

Reply:
I think PODIO design is in wrong here, if they followed single responsibility rules, the implementation of collections would be an overstep. If using PODIO collections for the return types aligns with our factory pattern, why can't we use  them in the method arguments as well?

My understanding was that the goal for algorithms was to encapsulate the logic of the algorithms, not the logic of the ePIC reconstruction. They would not be able to absorb all of it anyways, because the arrangement of factories into a pipeline is made by EICrecon.

Reply:
> why can't we use them in the method arguments as well?

We normally do; at least I do for all dRICH algorithms. This algorithm cannot, however, since its input `eicrecon::TrackingResultTrajectory` datatype is not an `EDM4*` datatype (which in my opinion should never have been allowed, since it can't be related to any other EDM objects).

> My understanding was that the goal for algorithms was to encapsulate the logic of the algorithms, not the logic of the ePIC reconstruction.

Correct, but any part of the ePIC software stack needs some "module" (model) to hold the data. This is one of the most important modules since it is the most difficult to change. Last summer, in the critical software decision process, the consensus for this model was PODIO+EDM4hep+EDM4eic and therefore we should expect to be able to use it and all of its features everywhere, including a "decoupled" algorithms framework.

This is a very good discussion, especially as input/feedback to the eventual decoupling of algorithms from the reconstruction framework; it's very non-trivial to draw the line between the algorithms and framework. However, I'm afraid this discussion has veered off topic from this PR review. Perhaps we can continue this in https://github.com/orgs/eic/discussions where there can be more visibility (even though it's been quiet nowadays...).

Reply:
> We normally do; at least I do for all dRICH algorithms.

I am satisfied with this answer. If you are saying that we can and should do this, we need to make a motion adopt this convention everywhere uniformly.

> This algorithm cannot, however, since its input eicrecon::TrackingResultTrajectory datatype is not an EDM4* datatype (which in my opinion should never have been allowed, since it can't be related to any other EDM objects).

Could you file an issue, so that we can transition it to EDM4eic?

---

> Correct, but any part of the ePIC software stack needs some "module" (model) to hold the data. This is one of the most important modules since it is the most difficult to change. Last summer, in the critical software decision process, the consensus for this model was PODIO+EDM4hep+EDM4eic and therefore we should expect to be able to use it and all of its features everywhere, including a "decoupled" algorithms framework.

I didn't think that this was the point of making that decision. The primary goal was setting the on-disk format PODIO, and picking a specific set of data structures (list structures; names and encodings of the fields) for algorithms was secondary. But what we are discussing is just about using collections API from PODIO in the algorithms.

I think, to have discussion here is very appropriate, as we have a concrete code to discuss. Hopefully, it doesn't disrupt your work or anything like that.

Reply:
> motion adopt this convention everywhere uniformly.

Yes, this should be brought up with those working on the `algorithms` API (@sly2j, whether algorithm input/output should be PODIO Collections or STL collections of (pointers to) PODIO objects). Fortunately EICrecon can handle both Collections and `std::vector<raw pointer>` (and probably other structures / smart pointers), and it's not too painful to refactor, should we force a convention on everyone. I like having conventions and consistency, but another solution would be to let algorithm developers choose.

However, I vote for PODIO collections, because `std::vector<raw pointer>` has wasted a lot of my time, especially causing difficulty when _reading_ EICrecon output with something else (successful writing does not guarantee successful reading downstream, so having CI benchmarks which _read_ the data is critical for stability).

> Could you file an issue, so that we can transition it to EDM4eic?

https://github.com/eic/EICrecon/issues/681

Reply:
In all cases podio collections: this is required to properly deal with associations and to allow for optimized memory layouts down the road (the collection and, by extension the frame owns the memory). There is much more that podio does behind the screens that is critical for good operations, especially in a multi-threaded context. We should under no circumstance work with raw pointers to Podio objects in our algorithms. `operator new` is banned in normal user code, and shared pointers/unique pointers should be used only for managing algorithm resources.  Note that many of the problems we have encountered in the last year were due to nonstandard implementation/usage of podio objects that could have been avoided by sticking to the API.